### DISCUSSION TITLE
Need Help in Publishing to a Client when it subscribes to a topic

### DISCUSSION THREAD

**Discussion by aqibamir** [moscajs/aedes#810] [Q&A]
Whenever a client subscribes to a topic , I have to deliver the latest data of the device to the subscribing client. In the following code whenever I use aedes.publish(), immediately after publishing the client gets disconnected. Is there something wrong with my approach ?

 ` aedes.authorizeSubscribe = async function (client, subscription, callback) {
    try {
    const splitted = subscription.topic.split('/');
 
    if (splitted.length < 4)
      callback(null, null);

    const projectDocument = await auth.verifyApiKey(fusionClient, splitted[0]);

    if (projectDocument == null) {
      return callback(new Error('Invlaid topic string'), null);
    }

    const res = await fusionClient.run('device', 'getDeviceData', [splitted[2], splitted[3], projectDocument.project._id]);
    
    if (res) {
      const returnPacket = {
        topic: subscription.topic,
        payload: Buffer.from(res.data),
        messageId: cuid(),
        qos: 2
      }
      aedes.publish(returnPacket, null);
    }
    return callback(new Error('DeviceID invalid'));
   }
   catch (err) {
    console.log(err);
    return callback(new Error(err));
   }
  }
 `

**Comment by robertsLando** [comment]
You should use `retain` flag for such purposes. What I suggest is to give a look at [MQTT foundamentals](https://www.hivemq.com/mqtt-essentials/) first

About your code you don't have to add messageId to the packet, just topic, payload, qos and retain flag
